src/server/pps/server/determined.go

```diff
@@ -0,0 +1,320 @@
+				assignRoleIds[r.Id] = struct{}{}
+				break
+			}
+		}
+	}
+	dwMap := make(map[int]detWorkspace)
+	for _, dw := range dwMap {
+		dwMap[dw.Id] = dw
+	}
+	for _, a := range resp.Assignments {
+		if _, ok := assignRoleIds[a.RoleId]; ok {
+			if a.ScopeCluster {
+				return nil
+			}
+			for _, id := range a.ScopeWorkspaceIds {
+				delete(dwMap, id)
+			}
+		}
+	}
+	if len(dwMap) > 0 {
+		var workspaces []string
+		for _, dw := range dwMap {
+			workspaces = append(workspaces, dw.Name)
+		}
+		return errors.Errorf("requested workspaces don't exist: %v", workspaces)
+	}
+	return nil
+}
+
+func (a *apiServer) provisionDeterminedPipelineUser(ctx context.Context, pi *pps.PipelineInfo, token string) (int, error) {
+	m := map[string]interface{}{
+		"username": pipelineUserName(pi.Pipeline),
+		"password": pi.Details.Determined.Password,
+	}
+	// TODO: handle existing existing user
+	data, err := a.determinedHttpRequest(ctx, "POST", "/api/v1/users", token, m)
+	if err != nil {
+		return 0, errors.Wrap(err, "provision determined user")
+	}
+	var resp detPostUserResponse
+	if err := json.Unmarshal(data, &resp); err != nil {
+		return 0, errors.Wrap(err, "unmarshal determined provision user response")
+	}
+	return resp.User.Id, nil
+}
+
+func (a *apiServer) assignDeterminedPipelineRole(ctx context.Context, token string, userId int, roleId int, workspaces []detWorkspace) error {
+	var roleAssignments []detUserRoleAssignment
+	for _, w := range workspaces {
+		roleAssignments = append(roleAssignments,
+			detUserRoleAssignment{
+				UserId: userId,
+				RoleAssignment: detRoleAssignment{
+					Role: detRoleAssignmentRole{
+						RoleId: roleId,
+					},
+					ScopeWorkspaceId: w.Id,
+				},
+			},
+		)
+	}
+	body := detUserRoleAssignmentsRequest{UserRoleAssignments: roleAssignments}
+	// craft request with workspace IDs
+	_, err := a.determinedHttpRequest(ctx, "POST", "/api/v1/roles/add-assignments", token, body)
+	if err != nil {
+		return errors.Wrap(err, "assign pipeline's determined user editor role")
+	}
+	return nil
+}
+
+func (a *apiServer) determinedHttpRequest(ctx context.Context, method, endpoint, token string, body any) ([]byte, error) {
+	var b *bytes.Buffer
+	if body != nil {
+		b = new(bytes.Buffer)
+		if err := json.NewEncoder(b).Encode(body); err != nil {
+			return nil, errors.Wrapf(err, "encode to json: %v", body)
+		}
+	}
+	req, err := http.NewRequestWithContext(ctx, method, a.env.Config.DeterminedURL+endpoint, b)
+	if err != nil {
+		return nil, errors.Wrap(err, "create determined http request")
+	}
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Authorization", "Bearer "+token)
+	client := &http.Client{Timeout: 10 * time.Second}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "execute determined http request")
+	}
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+		defer resp.Body.Close()
```

Comment:
I think the body always needs to be closed, even if it's a 500 internal server error or something.  Move this out of the if condition and always do it after Do returns a nil error.